### PR TITLE
[Merged by Bors] - feat: port Data/Sigma/Basic

### PR DESCRIPTION
**Pull request by kbuzzard**
My first attempt to port a file to mathlib4. Please review my code style carefully -- I'm still not 100% sure about the rules yet.

Notes: 

1) The file uses `Function.surjective.forall` once, and this declaration is in Mathlib.Logic.Function.Basic. Conversely, importing that file gives us everything we need. As a result the imports in this file look totally unrelated to the corresponding imports in the mathlib3 file.

2) The proof of `mk.inj_iff` in mathlib3 was `by simp` but this failed in mathlib4 so I knocked up a longer proof (lines 56-57). Should I worry about this? In Lean 3 `simp?` suggested `simp only [implies_true_iff, and.congr_left_iff]`, which didn't work in Lean 3, and `squeeze_simp` suggested `simp only` which (to my mild surprise) does ;-)

3) I also edited the module docstring because I found the Lean 3 docstring claim that a dependent product was "like `α × β`" and then immediately "a generalization of `α ⊕ β`" confusing (these are not the same alpha and beta; the idea is that the first alpha is bool in the second example; hopefully my explanation is now better).

### CODE REVIEW

File: Mathlib/Data/Sigma/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+# Sigma types
+
+This file proves basic results about sigma types.
+
+A sigma type is a dependent pair type. Like `α × β` but where the type of the second component
+depends on the first component. More precisely, given `β : ι → Type*`, `Sigma β` is made of stuff
+which is of type `β i` for some `i : ι`, so the sigma type is a disjoint union of types.
+For example, the sum type `X ⊕ Y` can be emulated using a sigma type, by taking `ι` with
+exactly two elements. See `equiv.sum_equiv_sigma_bool` (remark: probably not yet ported
+to Lean 4).
```
Comment (semorrison) [inline Mathlib/Data/Sigma/Basic.lean:20]:
```suggestion
exactly two elements. See `equiv.sum_equiv_sigma_bool`.
```
I think there's no need to leave warning about stuff we plan to port in future. These will probably never get cleaned up afterwards.

File: Mathlib/Data/Sigma/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+
+`Σ x, A x` is notation for `Sigma A` (note that this is `\Sigma`, not the sum operator `∑`).
+`Σ x y z ..., A x y z ...` is notation for `Σ x, Σ y, Σ z, ..., A x y z ...`. Here we have
+`α : Type*`, `β : α → Type*`, `γ : Π a : α, β a → Type*`, ...,
+`A : Π (a : α) (b : β a) (c : γ a b) ..., Type*`  with `x : α` `y : β x`, `z : γ x y`, ...
```
Comment (semorrison) [inline Mathlib/Data/Sigma/Basic.lean:25]:
```suggestion
`α : Type _`, `β : α → Type _`, `γ : Π a : α, β a → Type _`, ...,
`A : Π (a : α) (b : β a) (c : γ a b) ..., Type _`  with `x : α` `y : β x`, `z : γ x y`, ...
```

File: Mathlib/Data/Sigma/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+
+## Notes
+
+The definition of `Sigma` takes values in `Type*`. This effectively forbids `Prop`- valued sigma
+types. To that effect, we have `PSigma`, which takes value in `Sort*` and carries a more complicated
+universe signature as a consequence.
```
Comment (semorrison) [inline Mathlib/Data/Sigma/Basic.lean:31]:
```suggestion
The definition of `Sigma` takes values in `Type _`. This effectively forbids `Prop`- valued sigma
types. To that effect, we have `PSigma`, which takes value in `Sort _` and carries a more complicated
universe signature as a consequence.
```

File: Mathlib/Data/Sigma/Basic.lean [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+theorem eta : ∀ x : Σa, β a, Sigma.mk x.1 x.2 = x
+  | ⟨_, _⟩ => rfl
+
+@[ext]
+theorem ext {x₀ x₁ : Sigma β} (h₀ : x₀.1 = x₁.1) (h₁ : HEq x₀.2 x₁.2) : x₀ = x₁ := by
+  cases x₀
+  cases x₁
+  cases h₀
+  cases h₁
+  rfl
```
Comment (semorrison) [inline Mathlib/Data/Sigma/Basic.lean:69]:
```suggestion
  cases x₀; cases x₁; cases h₀; cases h₁; rfl
```
is fine, I think.

**Comment by semorrison** [comment]
Otherwise looks promising.

The linter is complaining about long lines.

The build is only failing because you need to add `import Mathlib.Data.Sigma.Basic` in the `Mathlib.lean` file in the root directory. (This is tedious, but apparently a good idea...)

File: Mathlib/Data/Sigma/Basic.lean (hunk 1 above)
Comment (kbuzzard) [reply Mathlib/Data/Sigma/Basic.lean:20]: I have punted for `Equiv.sum_equiv_sigma_bool` but I still don't properly understand naming conventions yet.

**Comment by kbuzzard** [comment]
I'm still a bit unclear about what to do next, by the way: should I mark the file as "done" in one or more places (e.g. do I need to make a PR to mathlib3 adding a comment in this file??), and furthermore do I do this now or after this is merged?

File: Mathlib/Data/Sigma/Basic.lean (hunk 1 above)
Comment (kbuzzard) [reply Mathlib/Data/Sigma/Basic.lean]: I've just looked up what mathlib3port says and have now punted for `Equiv.sumEquivSigmaBool`

**Comment by javra** [comment]
I think you should do it now at https://github.com/leanprover-community/mathlib/wiki/mathlib4-port-status, to prevent someone else to start on it.

**Comment by kbuzzard** [comment]
How the heck do I work out the mathlib commit sha? I got the file from mathlib3port.

**Comment by semorrison** [comment]
> How the heck do I work out the mathlib commit sha? I got the file from mathlib3port.

I've just updated the [porting wiki page](https://github.com/leanprover-community/mathlib4/wiki) to explain this. Obviously this is too painful for now, and we will need some CI assistance going forward.

---

8. Once it is accepted
  * Go to the [port status](https://github.com/leanprover-community/mathlib/wiki/mathlib4-port-status/_edit) page and change `No` to `Yes mathlib4#XXX`, to indicate that has been ported. This will update the results that `scripts/port_status.py` produces.
  * Make a PR to `mathlib` adding to the top of the module doc the text 
```text
> THIS FILE IS SYNCHRONIZED WITH MATHLIB4.
> https://github.com/leanprover-community/mathlib4/pull/XXX
> Any changes to this file require a corresponding PR to mathlib4.
```
  * Open up a current copy of mathlib, and do a once over to check that the new content in mathlib4 faithfully reflects the current content in mathlib. Then edit the [port status](https://github.com/leanprover-community/mathlib/wiki/mathlib4-port-status/_edit) page to say `Yes mathlib4#XXX YYYYYY`, where `YYYYYY` is the git commit sha for the version of `mathlib` you checked against (e.g. from `git log`).

**Comment by semorrison** [comment]
bors d+

**Comment by kbuzzard** [comment]
Thanks for all the help!

bors r+

**Pull request closed without merging**